playgame.py

```diff
@@ -5,55 +5,111 @@
 from tools.hanabi_card import HanabiColor
 import sys
 import logging
+from logging.handlers import RotatingFileHandler
+import os
 from tools.hanabi_moves import \
     HanabiDiscardAction, \
     HanabiPlayAction, \
     HanabiColorDiscloseAction, \
     HanabiRankDiscloseAction
 
-logger = logging.getLogger(__name__)
-
 def main(argv):
     args = parse_args()
-    prep_logger(args.log_dir, args.verbose, args.log_stderr)
+    logger = prep_logger(args.log_dir, args.verbose, args.log_stderr, len(args.players))
+    if args.is_tournament:
+        run_tournament(args, logger)
+    else:
+        run_one_game(args, logger)
+
+def generate_pairings(players):
+    res = []
+    for i in range(len(players)):
+        for j in range(i + 1, len(players)):
+            res.append([players[i], players[j]])
+    return res
+
+def run_tournament(args, logger):
+    tournament_scores = dict.fromkeys(args.players, 0)
+    pairings = generate_pairings(args.players)
+    for player1, player2 in pairings:
+        try:
+            game = HanabiGame([player1, player2], args.seed, args.variant)
+            game.play_game(args, logger)
+            score = game.table.score()
+        except:
+            #if a player has messed up, penalize both
+            #could diqualify the failing bot if we can determine who failed
+            logger.warning("{player1} or {player2} failed to complete a game".format(
+                player1 = player1,
+                player2 = player2
+                ))
+            score = -25
+        finally:
+            for handler in logger.handlers:
+                if handler.__class__ is RotatingFileHandler:
+                    handler.doRollover()
+            tournament_scores[player1] += score
+            tournament_scores[player2] += score
+    tournament_scores = {k: v / len(pairings) for k, v in tournament_scores.iteritems()}
```

Comment:
What calculation is happening here? It seems like it's sort of like an average, but the denominator is the total number of pairings, rather than only those including the player whose score is being used.

Do you think there's any value in considering other types of measurements? Is a bot with a high average but also high standard deviation better than one with a lower average but also lower std dev?

Reply:
Yes, this is the most naive average. I have issue #3 that still isn't really addressed...